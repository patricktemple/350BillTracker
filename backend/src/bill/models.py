import enum
from uuid import uuid4

import flask
from sqlalchemy import Boolean, Column, Enum, ForeignKey, Integer, Text, sql
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import foreign, relationship, remote

from ..models import TIMESTAMP, UUID, db
from ..user.models import User
from ..utils import now

DEFAULT_TWITTER_SEARCH_TERMS = [
    "solar",
    "climate",
    "wind power",
    "renewable",
    "fossil fuel",
]


class StateChamber(enum.Enum):
    SENATE = 1
    ASSEMBLY = 2


class Bill(db.Model):
    """
    Base table for all bills, both city and state. Contains any info that's
    relevant for both kinds of bills.

    Bills are polymorphic. This table has a "type" which specifies whether it's
    a city or a state bill, and there will be an associated row in the CityBill
    and StateBill with more info on it."""

    __tablename__ = "bills"

    class BillType(enum.Enum):
        CITY = 1
        STATE = 2

    id = Column(UUID, primary_key=True, default=uuid4)

    # For city this is the "MatterName", for state it is the "title"
    name = Column(Text, nullable=False)

    # For city this is the "title", for state this is the "summary"
    description = Column(Text, nullable=False)

    # Info on child objects:
    type = Column(Enum(BillType), nullable=False)

    # If this is is a city bill, points to more specific data on that.
    city_bill = relationship(
        "CityBill",
        back_populates="bill",
        uselist=False,
        cascade="all, delete",
        lazy="joined",
    )

    # If this is is a state bill, points to more specific data on that.
    state_bill = relationship(
        "StateBill",
        back_populates="bill",
        uselist=False,
        cascade="all, delete",
        lazy="joined",
    )

    # Data we track
    notes = Column(Text, nullable=False, server_default="")
    nickname = Column(Text, nullable=False, server_default="")
    twitter_search_terms = Column(
        ARRAY(Text), nullable=False, default=DEFAULT_TWITTER_SEARCH_TERMS
    )

    power_hours = relationship(
        "PowerHour", back_populates="bill", cascade="all, delete"
    )
    attachments = relationship(
        "BillAttachment", back_populates="bill", cascade="all, delete"
    )

    @property
    def display_name(self):
        return self.nickname if self.nickname else self.name

    @property
    def tracked(self):
        return True

    @property
    def status(self):
        if self.type == Bill.BillType.CITY:
            return self.city_bill.status

        senate_status = (
            self.state_bill.senate_bill.status
            if self.state_bill.senate_bill
            else "(No Senate bill)"
        )
        assembly_status = (
            self.state_bill.assembly_bill.status
            if self.state_bill.assembly_bill
            else "(No Assembly bill)"
        )

        return f"{senate_status} / {assembly_status}"

    @property
    def code_name(self):
        if self.type == Bill.BillType.CITY:
            return self.city_bill.file

        senate_print_no = (
            self.state_bill.senate_bill.base_print_no
            if self.state_bill.senate_bill
            else "(No Senate bill)"
        )
        assembly_print_no = (
            self.state_bill.assembly_bill.base_print_no
            if self.state_bill.assembly_bill
            else "(No Assembly bill)"
        )

        return f"{senate_print_no} / {assembly_print_no} from {self.state_bill.session_year} session"
    
    user_bill_settings = relationship("UserBillSettings", back_populates="bill", cascade="all, delete-orphan")


class BillAttachment(db.Model):
    __tablename__ = "bill_attachments"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(UUID, ForeignKey("bills.id"), nullable=False, index=True)
    bill = relationship("Bill", back_populates="attachments")

    name = Column(Text)
    url = Column(Text, nullable=False)


class PowerHour(db.Model):
    __tablename__ = "power_hours"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(UUID, ForeignKey("bills.id"), nullable=False, index=True)

    title = Column(Text)
    spreadsheet_url = Column(Text, nullable=False)
    spreadsheet_id = Column(Text, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)
    bill = relationship("Bill", back_populates="power_hours")


class CityBill(db.Model):
    """
    City-specific details about an NYC bill. There must also be an associated
    row in the Bill parent table that has more general bill info (like name).
    """

    __tablename__ = "city_bills"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)

    # ID in the City Council API
    city_bill_id = Column(Integer, nullable=False, unique=True, index=True)

    # Code name often used to discuss the bill, like Int 2317-2021
    file = Column(Text, nullable=False)

    # The parent Bill object that represents this.
    bill = relationship("Bill", back_populates="city_bill", lazy="joined")

    intro_date = Column(TIMESTAMP, nullable=False)

    status = Column(Text, nullable=False)

    # Which amendment version is currently active. We only track one at a time,
    # and sponsorships are tied to that particular version.
    active_version = Column(Text, nullable=False)

    sponsorships = relationship(
        "CitySponsorship", back_populates="city_bill", cascade="all, delete"
    )

    # Committee name
    council_body = Column(Text)


# TODO: Understand bill substitution and change if needed
class StateBill(db.Model):
    """
    State-specific details about a state bill. There must also be an associated
    row in the Bill parent table that has more general bill info (like name)."""

    __tablename__ = "state_bills"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)
    bill = relationship(Bill, back_populates="state_bill", lazy="joined")

    # The start of the 2-year legislative session this belongs to.
    session_year = Column(Integer, nullable=False)

    senate_bill = relationship(
        "SenateBill",
        back_populates="state_bill",
        uselist=False,
        cascade="all, delete",
    )
    assembly_bill = relationship(
        "AssemblyBill",
        back_populates="state_bill",
        uselist=False,
        cascade="all, delete",
    )


class StateChamberMixin:
    """
    Mixin for a bill in either the Senate or Assembly chamber. The have
    identical structure but are kept in separate tables."""

    active_version = Column(Text, nullable=False)
    status = Column(Text, nullable=False)
    base_print_no = Column(Text, nullable=False)

    @declared_attr
    def bill_id(self):
        return Column(UUID, ForeignKey(StateBill.bill_id), primary_key=True)

    # The Senate and the Assembly each run separate websites, and each website can
    # lookup both senate and assembly bills. So they're redundant websites with
    # very different UI. Therefore senate and assembly bills each have a
    # both assembly and senate websites.
    @property
    def senate_website(self):
        return f"https://www.nysenate.gov/legislation/bills/{self.state_bill.session_year}/{self.base_print_no}"

    @property
    def assembly_website(self):
        return f"https://nyassembly.gov/leg/?term={self.state_bill.session_year}&bn={self.base_print_no}"


class SenateBill(db.Model, StateChamberMixin):
    __tablename__ = "senate_bills"

    state_bill = relationship(StateBill, back_populates="senate_bill")
    sponsorships = relationship(
        "SenateSponsorship",
        back_populates="chamber_bill",
        cascade="all, delete-orphan",
    )


class AssemblyBill(db.Model, StateChamberMixin):
    __tablename__ = "assembly_bills"

    state_bill = relationship(StateBill, back_populates="assembly_bill")
    sponsorships = relationship(
        "AssemblySponsorship",
        back_populates="chamber_bill",
        cascade="all, delete-orphan",
    )


class UserBillSettings(db.Model):
    __tablename__ = "user_bill_settings"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)
    user_id = Column(UUID, ForeignKey(User.id), primary_key=True)

    send_bill_update_notifications = Column(
        Boolean, nullable=False, server_default=sql.false()
    )

    user = relationship(User, lazy="joined", back_populates="bill_settings")
    bill = relationship(Bill, back_populates="user_bill_settings")


Bill.viewer_settings = relationship(
    UserBillSettings,
    primaryjoin=(remote(UserBillSettings.bill_id) == foreign(Bill.id))
    & (
        UserBillSettings.user_id
        == sql.bindparam(
            "request_user_id", callable_=lambda: flask.g.request_user_id
        )
    ),
    viewonly=True,
)
