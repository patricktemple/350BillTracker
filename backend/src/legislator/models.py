from uuid import uuid4

from sqlalchemy import Column, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from ..models import TIMESTAMP, UUID, db


class Legislator(db.Model):
    __tablename__ = "legislators"

    # These come from the API
    id = Column(Integer, primary_key=True)
    name = Column(Text, nullable=False)
    term_start = Column(TIMESTAMP)
    term_end = Column(TIMESTAMP)
    email = Column(Text)
    district_phone = Column(Text)
    legislative_phone = Column(Text)

    sponsorships = relationship("BillSponsorship", back_populates="legislator")
    staffers = relationship("Staffer", back_populates="legislator")

    borough = Column(Text)
    website = Column(Text)

    # These are added by our static data
    twitter = Column(Text)  # exclude the @ symbol
    party = Column(Text)

    # Track our own info on the bill.
    notes = Column(Text)

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def twitter_url(self):
        return (
            f"https://www.twitter.com/{self.twitter}" if self.twitter else None
        )


class Staffer(db.Model):
    __tablename__ = "staffers"

    id = Column(UUID, primary_key=True, default=uuid4)
    name = Column(Text, nullable=False)
    title = Column(Text)
    email = Column(Text)
    phone = Column(Text)
    legislator_id = Column(
        Integer, ForeignKey("legislators.id"), nullable=False
    )
    twitter = Column(Text)

    legislator = relationship("Legislator", back_populates="staffers")

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def display_string(self):
        contact_methods = [self.phone, self.email, self.display_twitter]
        contact_methods = [c for c in contact_methods if c]
        contact_string = ", ".join(contact_methods)
        if not contact_string:
            contact_string = "No contact info"
        title_string = f"{self.title} - " if self.title else ""

        return f"{title_string}{self.name} ({contact_string})"
