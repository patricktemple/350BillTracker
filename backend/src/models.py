from uuid import uuid4

from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    ForeignKey,
    Integer,
    Text,
    TypeDecorator,
    sql,
)
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.dialects.postgresql import TIMESTAMP as _TIMESTAMP
from sqlalchemy.dialects.postgresql import UUID as _UUID
from sqlalchemy.orm import relationship

from .app import app
from .utils import now

db = SQLAlchemy(app)


class UUID(TypeDecorator):
    impl = _UUID(as_uuid=True)


class TIMESTAMP(TypeDecorator):
    impl = _TIMESTAMP(timezone=True)


# TODO: UUIDs for some PKs?
class BillAttachment(db.Model):
    __tablename__ = "bill_attachments"

    id = Column(Integer, primary_key=True)
    bill_id = Column(
        Integer, ForeignKey("bills.id"), nullable=False, index=True
    )
    bill = relationship("Bill", back_populates="attachments")

    name = Column(Text)
    url = Column(Text, nullable=False)


class PowerHour(db.Model):
    __tablename__ = "power_hours"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(
        Integer, ForeignKey("bills.id"), nullable=False, index=True
    )

    title = Column(Text)
    spreadsheet_url = Column(Text, nullable=False)
    spreadsheet_id = Column(Text, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)
    bill = relationship("Bill", back_populates="power_hours")
