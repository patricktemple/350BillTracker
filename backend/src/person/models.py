import enum
from uuid import uuid4

from sqlalchemy import Column, Enum, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from ..models import TIMESTAMP, UUID, db


class Person(db.Model):
    """
    Base table for every kind of person we track. General info shared by all
    people, such as name and phone, are kept here.

    A person is polymorphic. There is also a "type" that specifies which kind
    of person this is (such as a senator), and there are
    separate tables for each type of person that has more specific details.
    """

    __tablename__ = "persons"

    class PersonType(enum.Enum):
        # City
        COUNCIL_MEMBER = 1

        # State
        ASSEMBLY_MEMBER = 2
        SENATOR = 3

        # General
        STAFFER = 4

    # This is the internal ID shared by all people
    id = Column(UUID, primary_key=True, default=uuid4)

    name = Column(Text, nullable=False)
    title = Column(Text)
    email = Column(Text)
    phone = Column(Text)

    twitter = Column(Text)  # excludes the @ symbol

    # Track our own info on the person
    notes = Column(Text)

    type = Column(Enum(PersonType), nullable=False)

    # If this person is a council member, points to the object that has more specific
    # details about them in that role.
    council_member = relationship(
        "CouncilMember", back_populates="person", uselist=False, lazy="joined"
    )

    # If this person is a state senator, points to the object that has more specific
    # details about them in that role.
    senator = relationship(
        "Senator", back_populates="person", uselist=False, lazy="joined"
    )

    # If this person is a state assembly, points to the object that has more specific
    # details about them in that role.
    assembly_member = relationship(
        "AssemblyMember", back_populates="person", uselist=False, lazy="joined"
    )

    office_contacts = relationship("OfficeContact", back_populates="person")

    # These are added by our static data
    # TODO: Make this an enum?
    party = Column(Text)

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def twitter_url(self):
        return (
            f"https://www.twitter.com/{self.twitter}" if self.twitter else None
        )


class CouncilMember(db.Model):
    """
    Data about a specific council member. For each row in this there must also
    be a Person that has more general info about this person, such as name.
    """

    __tablename__ = "council_members"

    # Foreign key to Person parent table
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    # ID of the City Council API object, which is also called Person
    city_council_person_id = Column(
        Integer, nullable=False, index=True, unique=True
    )

    term_start = Column(TIMESTAMP)
    term_end = Column(TIMESTAMP)

    # District phone is stored in Person.phone
    legislative_phone = Column(Text)

    sponsorships = relationship(
        "CitySponsorship", back_populates="council_member"
    )

    # The parent Person object.
    person = relationship(
        "Person", back_populates="council_member", lazy="joined"
    )

    borough = Column(Text)
    website = Column(Text)


class Senator(db.Model):
    """
    Data about a specific state senator. For each row in this there must also
    be a Person that has more general info about this person, such as name.
    """

    __tablename__ = "senators"

    # Foreign key to Person parent table
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    # The parent Person object.
    person = relationship("Person", back_populates="senator", lazy="joined")

    sponsorships = relationship("SenateSponsorship", back_populates="senator")

    state_member_id = Column(Integer, nullable=False, unique=True)

    district = Column(Integer)

    @property
    def website(self):
        return (
            f"https://www.nysenate.gov/district/{self.district}"
            if self.district
            else None
        )


class AssemblyMember(db.Model):
    """
    Data about a specific state assembly member. For each row in this there must
    also be a Person that has more general info about this person, such as name.
    """

    __tablename__ = "assembly_members"

    # Foreign key to Person parent table
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    # The parent Person object.
    person = relationship(
        "Person", back_populates="assembly_member", lazy="joined"
    )

    state_member_id = Column(Integer, nullable=False, unique=True)

    sponsorships = relationship(
        "AssemblySponsorship", back_populates="assembly_member"
    )

    district = Column(Integer)

    @property
    def website(self):
        return (
            f"https://nyassembly.gov/mem/?ad={self.district}"
            if self.district
            else None
        )


# Staffers have a single boss. They're one to many.
class Staffer(db.Model):
    __tablename__ = "staffers"

    # ID of the person themselves:
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    boss_id = Column(UUID, ForeignKey(Person.id), nullable=False, index=True)

    # The Person object that represents general info about this person (NOT the person that
    # they work for.)
    person = relationship(
        "Person",
        foreign_keys=[person_id],
        back_populates="staffer",
        lazy="joined",
    )

    # The Person that this staffer works for.
    boss = relationship("Person", foreign_keys=[boss_id])


# If this person is a Staffer, gets the Staffer object that represents them.
Person.staffer = relationship(
    Staffer,
    foreign_keys=[Staffer.person_id],
    back_populates="person",
    cascade="all, delete",
    uselist=False,
)

# Gets all the Persons that work for this person as staffers.
Person.staffer_persons = relationship(
    Person,
    secondary="staffers",
    primaryjoin=Person.id == Staffer.boss_id,
    secondaryjoin=Staffer.person_id == Person.id,
    viewonly=True,
)


class OfficeContact(db.Model):
    __tablename__ = "office_contacts"

    id = Column(UUID, primary_key=True, default=uuid4)

    person_id = Column(UUID, ForeignKey(Person.id), index=True)

    # address type enum, maybe?

    phone = Column(Text)
    fax = Column(Text)
    city = Column(Text)

    person = relationship(
        "Person", back_populates="office_contacts", lazy="joined"
    )