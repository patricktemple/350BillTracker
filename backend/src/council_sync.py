import logging
from datetime import datetime, timezone

from requests import HTTPError

from .bill.models import Bill, CityBill
from .council_api import (
    get_bill_sponsors,
    get_current_council_members,
    get_person,
    lookup_bill,
)
from .models import db
from .person.models import CouncilMember, Person
from .sponsorship.models import CitySponsorship
from .static_data import COUNCIL_DATA_BY_LEGISLATOR_ID
from .utils import now


def add_council_members():
    """Adds basic information about each current council member based on their
    OfficeRecord object in the API. This is missing key fields that need
    to be filled in with other sources."""
    members = get_current_council_members()

    for member in members:
        city_council_person_id = member["OfficeRecordPersonId"]

        existing_council_member = CouncilMember.query.filter_by(
            city_council_person_id=city_council_person_id
        ).one_or_none()
        if existing_council_member:
            council_member = existing_council_member
            person = existing_council_member.person
        else:
            person = Person(type=Person.PersonType.COUNCIL_MEMBER)
            person.council_member = CouncilMember()
            council_member = person.council_member
            db.session.add(person)

        person.name = member["OfficeRecordFullName"]
        person.title = "City Council Member"  # TODO
        council_member.city_council_person_id = city_council_person_id
        council_member.term_start = datetime.fromisoformat(
            member["OfficeRecordStartDate"]
        ).replace(tzinfo=timezone.utc)
        council_member.term_end = datetime.fromisoformat(
            member["OfficeRecordEndDate"]
        ).replace(tzinfo=timezone.utc)

    db.session.commit()


def fill_council_person_data_from_api():
    """For all council members in the DB, updates their contact info and other details
    based on the Person API and our own static data.
    """
    council_members = CouncilMember.query.all()

    # TODO make sure the query joined loads the persons
    for council_member in council_members:
        try:
            data = get_person(council_member.city_council_person_id)
            council_member.person.email = data["PersonEmail"]
            council_member.person.phone = data["PersonPhone"]
            council_member.legislative_phone = data["PersonPhone2"]
            council_member.website = data["PersonWWW"]
            # Borough exists here but we prefer the cleaned static data
        except HTTPError:
            logging.exception(
                f"Could not get Person {council_member.city_council_person_id} from API"
            )
            continue

    db.session.commit()


def fill_council_person_static_data():
    council_members = CouncilMember.query.all()

    for council_member in council_members:
        member_data = COUNCIL_DATA_BY_LEGISLATOR_ID.get(
            council_member.city_council_person_id
        )
        if not member_data:
            logging.warning(
                f"Found a legislator without static data: {council_member.city_council_person_id} {council_member.person.name}"
            )
        else:
            council_member.person.twitter = member_data["twitter"]
            council_member.person.party = member_data["party"]

            # Name and borough both exist in the API but the static data has a
            # cleaned-up version.
            council_member.person.name = member_data["name"]
            council_member.borough = member_data["borough"]

    legislator_ids_from_db = set(
        [l.city_council_person_id for l in council_members]
    )
    if diff := set(COUNCIL_DATA_BY_LEGISLATOR_ID.keys()).difference(
        legislator_ids_from_db
    ):
        unmatched_static_data = [
            COUNCIL_DATA_BY_LEGISLATOR_ID[id] for id in diff
        ]
        logging.warning(
            f"Static data has some legislators not in the DB: {unmatched_static_data}"
        )

    db.session.commit()


# Bills ----------------------------------------------------------------------


def _update_bill(bill):
    bill_data = lookup_bill(bill.city_bill.city_bill_id)
    logging.info(
        f"Updating bill {bill.city_bill.city_bill_id} and got {bill_data}"
    )

    bill.name = bill_data["name"]
    for key in bill_data["city_bill"].keys():
        setattr(bill.city_bill, key, bill_data["city_bill"][key])


def sync_bill_updates():
    bills = Bill.query.filter_by(type=Bill.BillType.CITY).all()

    for bill in bills:
        _update_bill(bill)
        db.session.commit()


def update_bill_sponsorships(city_bill, set_added_at=False):
    """
    Updates sponsorships for a given bill:
    1) Deletes any previous sponsorships
    2) Adds all new sponsorships
    3) If new sponsors aren't in the existing legislators (e.g. they're not longer in office),
       ignore them.
    """
    previous_bill_sponsorships = CitySponsorship.query.filter_by(
        bill_id=city_bill.bill_id
    ).all()
    previous_bill_sponsorships_by_city_id = {
        s.council_member.city_council_person_id: s
        for s in previous_bill_sponsorships
    }

    updated_bill_sponsorships = get_bill_sponsors(city_bill.city_bill_id)

    council_members_for_updated_sponsorships = CouncilMember.query.filter(
        CouncilMember.city_council_person_id.in_(
            [s["MatterSponsorNameId"] for s in updated_bill_sponsorships]
        )
    ).all()
    council_members_for_updated_sponsorships_by_city_id = {
        c.city_council_person_id: c
        for c in council_members_for_updated_sponsorships
    }

    # THIS is super confusing, rewrite it and simplify
    for sponsorship in updated_bill_sponsorships:
        council_member_person_id = sponsorship["MatterSponsorNameId"]

        if (
            council_member_person_id
            not in council_members_for_updated_sponsorships_by_city_id
        ):
            # Can't insert the sponsorship without its foreign key object
            # TODO: Instead, insert a stub for them or something
            logging.warning(
                f"Did not find legislator {council_member_person_id} in db, ignoring..."
            )
            continue

        sponsorship = previous_bill_sponsorships_by_city_id.get(
            council_member_person_id
        )
        if sponsorship:
            # Remove sponsors from this set until we're left with only those
            # sponsorships that were rescinded recently.
            del previous_bill_sponsorships_by_city_id[council_member_person_id]
        else:
            sponsorship = CitySponsorship(
                bill_id=city_bill.bill_id,
                council_member_id=council_members_for_updated_sponsorships_by_city_id[
                    council_member_person_id
                ].person_id,
                sponsor_sequence=sponsorship["MatterSponsorSequence"],
            )
            if set_added_at:
                sponsorship.added_at = now()

            db.session.add(sponsorship)

    for lost_sponsor in previous_bill_sponsorships_by_city_id.values():
        db.session.delete(lost_sponsor)


def update_all_sponsorships():
    bills = Bill.query.all()
    for bill in bills:
        logging.info(f"Updating sponsorships for bill {bill.id} {bill.name}")
        update_bill_sponsorships(bill.city_bill, set_added_at=True)
        db.session.commit()
