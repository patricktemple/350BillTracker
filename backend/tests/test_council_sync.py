from datetime import datetime, timezone
from uuid import uuid4

import responses
from freezegun import freeze_time

from src.bill.models import Bill, CityBill
from src.council_sync import (add_council_members,
                              fill_council_person_data_from_api,
                              fill_council_person_static_data,
                              sync_bill_updates, update_all_sponsorships)
from src.models import db
from src.person.models import CouncilMember, Person
from src.sponsorship.models import CitySponsorship
from src.static_data import COUNCIL_DATA_BY_LEGISLATOR_ID


@responses.activate
@freeze_time("2021-10-12")
def test_add_council_members():
    existing_person = Person(
        name="Corey Johnson", type=Person.PersonType.COUNCIL_MEMBER
    )
    existing_person.council_member = CouncilMember(
        term_start="2000-01-01", city_council_person_id=1
    )
    db.session.add(existing_person)

    responses.add(
        responses.GET,
        url="https://webapi.legistar.com/v1/nyc/officerecords?token=fake_token&%24filter=OfficeRecordBodyName+eq+%27City+Council%27+and+OfficeRecordStartDate+le+datetime%272021-10-12%27+and+OfficeRecordEndDate+ge+datetime%272021-10-12%27",
        json=[
            {
                "OfficeRecordFullName": "Corey Johnson the 2nd",
                "OfficeRecordPersonId": "1",
                "OfficeRecordStartDate": "2021-01-01",
                "OfficeRecordEndDate": "2022-01-01",
            },
            {
                "OfficeRecordFullName": "Brad Lander",
                "OfficeRecordPersonId": "2",
                "OfficeRecordStartDate": "2021-01-01",
                "OfficeRecordEndDate": "2022-01-01",
            },
        ],
    )

    add_council_members()

    assert Person.query.count() == 2
    assert CouncilMember.query.count() == 2

    corey_johnson = CouncilMember.query.filter_by(
        city_council_person_id=1
    ).one()
    assert corey_johnson.person.name == "Corey Johnson the 2nd"
    assert corey_johnson.term_start == datetime(
        2021, 1, 1, tzinfo=timezone.utc
    )
    assert corey_johnson.term_end == datetime(2022, 1, 1, tzinfo=timezone.utc)

    brad_lander = CouncilMember.query.filter_by(city_council_person_id=2).one()
    assert brad_lander.person.name == "Brad Lander"


@responses.activate
def test_fill_council_person_data():
    person_to_update = Person(
        name="Corey Johnson", type=Person.PersonType.COUNCIL_MEMBER
    )
    person_to_update.council_member = CouncilMember(city_council_person_id=1)
    db.session.add(person_to_update)

    person_without_new_data = Person(
        name="Person who was impeached and removed",
        type=Person.PersonType.COUNCIL_MEMBER,
    )
    person_without_new_data.council_member = CouncilMember(
        city_council_person_id=2
    )
    db.session.add(person_without_new_data)

    responses.add(
        responses.GET,
        url="https://webapi.legistar.com/v1/nyc/persons/1?token=fake_token",
        json={
            "PersonEmail": "corey@council.nyc.gov",
            "PersonPhone": "555-111-1111",
            "PersonPhone2": "888-888-8888",
            "PersonCity1": "New York",
            "PersonWWW": "https://www.example.com/",
        },
    )

    # We just want to check that a 404 doesn't crash the whole thing
    responses.add(
        responses.GET,
        url="https://webapi.legistar.com/v1/nyc/persons/2?token=fake_token",
        status=404,
    )

    fill_council_person_data_from_api()

    assert Person.query.count() == 2
    assert CouncilMember.query.count() == 2

    corey = CouncilMember.query.filter_by(city_council_person_id=1).one()
    assert corey.person.email == "corey@council.nyc.gov"
    assert corey.person.phone == "555-111-1111"
    assert corey.legislative_phone == "888-888-8888"
    assert corey.website == "https://www.example.com/"


def test_fill_council_person_static_data():
    corey_static = COUNCIL_DATA_BY_LEGISLATOR_ID[7631]
    person_to_update = Person(
        name="Corey Johnson badly formatted name----",
        email="existing-email@council.nyc.gov",
        type=Person.PersonType.COUNCIL_MEMBER,
    )
    person_to_update.council_member = CouncilMember(
        city_council_person_id=7631
    )
    db.session.add(person_to_update)

    person_without_static_data = Person(
        name="Person without static data",
        type=Person.PersonType.COUNCIL_MEMBER,
    )
    person_without_static_data.council_member = CouncilMember(
        city_council_person_id=2,
    )
    db.session.add(person_without_static_data)

    fill_council_person_static_data()

    corey = CouncilMember.query.filter_by(city_council_person_id=7631).one()

    assert corey.person.name == corey_static["name"]
    assert corey.person.twitter == corey_static["twitter"]
    assert corey.person.party == corey_static["party"]
    assert corey.borough == corey_static["borough"]
    assert corey.person.email == "existing-email@council.nyc.gov"


@responses.activate
def test_sync_bill_updates(bill):
    responses.add(
        responses.GET,
        url="https://webapi.legistar.com/v1/nyc/matters/1?token=fake_token",
        json={
            "MatterId": bill.city_bill.city_bill_id,
            "MatterFile": "New file",
            "MatterName": "New name",
            "MatterTitle": "New title",
            "MatterBodyName": "New body",
            "MatterIntroDate": "2021-01-01T00:00:00",
            "MatterStatusName": "New status",
            "MatterVersion": "New version",
        },
    )

    sync_bill_updates()

    result = Bill.query.one()
    assert result.name == "New name"
    assert result.description == "New title"
    assert result.city_bill.file == "New file"
    assert result.city_bill.council_body == "New body"
    assert result.city_bill.status == "New status"
    assert result.city_bill.intro_date == datetime(
        2021, 1, 1, tzinfo=timezone.utc
    )
    assert result.nickname == bill.nickname
    assert result.city_bill.active_version == "New version"


@responses.activate
@freeze_time("2021-1-1")
def test_update_sponsorships__new_sponsor(bill):
    person = Person(name="Patrick", type=Person.PersonType.COUNCIL_MEMBER)
    person.council_member = CouncilMember(city_council_person_id=1)
    db.session.add(person)

    wrong_version_person = Person(
        name="Jim", type=Person.PersonType.COUNCIL_MEMBER
    )
    wrong_version_person.council_member = CouncilMember(
        city_council_person_id=2
    )
    db.session.add(wrong_version_person)

    responses.add(
        responses.GET,
        url=f"https://webapi.legistar.com/v1/nyc/matters/{bill.city_bill.city_bill_id}/sponsors?token=fake_token",
        json=[
            {
                "MatterSponsorNameId": 1,
                "MatterSponsorSequence": 0,
                "MatterSponsorMatterVersion": "A",
            },
            {
                # This one isn't found in the DB
                "MatterSponsorNameId": 3,
                "MatterSponsorSequence": 1,
                "MatterSponsorMatterVersion": "A",
            },
            {
                "MatterSponsorNameId": 2,
                "MatterSponsorSequence": 0,
                "MatterSponsorMatterVersion": "different",
            },
        ],
    )

    update_all_sponsorships()

    sponsorship = CitySponsorship.query.one()
    assert sponsorship.council_member_id == person.id
    assert sponsorship.bill_id == bill.id
    assert sponsorship.added_at == datetime(2021, 1, 1, tzinfo=timezone.utc)


@responses.activate
@freeze_time("2021-1-1")
def test_update_sponsorships__sponsorship_already_exists(bill):
    person = Person(
        id=uuid4(), name="Patrick", type=Person.PersonType.COUNCIL_MEMBER
    )
    person.council_member = CouncilMember(city_council_person_id=1)
    db.session.add(person)

    sponsorship = CitySponsorship(
        bill_id=bill.id,
        council_member_id=person.id,
        added_at=datetime(2000, 1, 1, tzinfo=timezone.utc),
        sponsor_sequence=0,
    )
    db.session.add(sponsorship)

    responses.add(
        responses.GET,
        url=f"https://webapi.legistar.com/v1/nyc/matters/{bill.city_bill.city_bill_id}/sponsors?token=fake_token",
        json=[
            {
                "MatterSponsorNameId": 1,
                "MatterSponsorSequence": 0,
                "MatterSponsorMatterVersion": "A",
            }
        ],
    )

    update_all_sponsorships()

    sponsorship = CitySponsorship.query.one()
    assert sponsorship.council_member_id == person.id
    assert sponsorship.bill_id == bill.id
    assert sponsorship.added_at == datetime(2000, 1, 1, tzinfo=timezone.utc)


@responses.activate
@freeze_time("2021-1-1")
def test_update_sponsorships__remove_sponsorship(bill):
    person = Person(
        id=uuid4(), name="Patrick", type=Person.PersonType.COUNCIL_MEMBER
    )
    person.council_member = CouncilMember(city_council_person_id=1)
    db.session.add(person)

    sponsorship = CitySponsorship(
        bill_id=bill.id,
        council_member_id=person.id,
        added_at=datetime(2000, 1, 1, tzinfo=timezone.utc),
        sponsor_sequence=0,
    )
    db.session.add(sponsorship)

    responses.add(
        responses.GET,
        url=f"https://webapi.legistar.com/v1/nyc/matters/{bill.city_bill.city_bill_id}/sponsors?token=fake_token",
        json=[],
    )

    update_all_sponsorships()

    assert CitySponsorship.query.count() == 0
