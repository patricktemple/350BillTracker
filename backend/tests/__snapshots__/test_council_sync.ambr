# name: test_add_council_committees
  <class 'list'> [
    <class 'tuple'> (
      'Existing committee',
      1,
      'Committee',
    ),
    <class 'tuple'> (
      'New valid committee',
      2,
      'Committee',
    ),
    <class 'tuple'> (
      'Valid subcommittee',
      3,
      'Subcommittee',
    ),
    <class 'tuple'> (
      'Land Use committee',
      4,
      'Land Use',
    ),
  ]
---
